File: 2426.cpp
/*
In this problem we have to find 0<=i<j<=n-1
and such i and j pairs where num1[i]-num1[j]<=num2[i]-num2[j]+diff

we can re-arrange the condition as
num1[i]-num2[i]<=num1[j]-num2[j]+diff

so first we simply have to create a difference vector

vector<int> difference;
for (int i=0;i<nums1.size();i++){
        difference.push_back(num1[i]-num2[i]);
}

and now we just have to find num1[i]-num2[i]<=num1[j]-num2[j]+diff

then we do merge sort (recursive) and also creae a count function incrementor




*/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    long long int count;
   
    void checkCount(vector<int>& nums, int start, int mid, int end, int d) {
        int l = start, r = mid + 1;
        while(l <= mid && r <= end) {
            if(nums[l] <= (nums[r]+d)) {
                count += (end - r + 1);
                l++;
            }
            else {
                r++;
            }
        }
        sort(nums.begin() + start, nums.begin() + end + 1);
        return;
    }
    
    void mergeSort(vector<int>& nums, int start, int end, int d) {
        if(start == end) 
            return;
        
        int mid = (start + end)/2;
        mergeSort(nums, start, mid, d);
        mergeSort(nums, mid+1, end, d);
        
        checkCount(nums, start, mid, end, d);
        return;
    }
    
    long long numberOfPairs(vector<int>& a, vector<int>& b, int d) {
        count = 0;
        int n = a.size();
        vector<int>c(n);  
        
        for(int i=0; i<n; i++)
            c[i] = a[i]-b[i];               
			
        mergeSort(c, 0, n-1, d);
		
        return count;
    }
};

int main() {
    Solution sol;
    vector<int> nums1 = {3, -1};
    vector<int> nums2 = {-2, 2}; 
    int diff = -1;
    
    cout << sol.numberOfPairs(nums1, nums2, diff);
    return 0;
}

File: 2426.rs
struct Solution {
    count: i64,
}

impl Solution {
    fn new() -> Self {
        Solution { count: 0 }
    }

    fn check_count(&mut self, nums: &mut Vec<i32>, start: usize, mid: usize, end: usize, d: i32) {
        let mut l = start;
        let mut r = mid + 1;
        
        while l <= mid && r <= end {
            if nums[l] <= nums[r] + d {
                self.count += (end - r + 1) as i64;
                l += 1;
            } else {
                r += 1;
            }
        }
        
        // Sort the slice from start to end
        nums[start..=end].sort_unstable();
    }

    fn merge_sort(&mut self, nums: &mut Vec<i32>, start: usize, end: usize, d: i32) {
        if start >= end {
            return;
        }
        
        let mid = start + (end - start) / 2;
        self.merge_sort(nums, start, mid, d);
        self.merge_sort(nums, mid + 1, end, d);
        
        self.check_count(nums, start, mid, end, d);
    }

    fn number_of_pairs(&mut self, a: Vec<i32>, b: Vec<i32>, d: i32) -> i64 {
        self.count = 0;
        let n = a.len();
        
        // Create difference array
        let mut c: Vec<i32> = a.iter()
            .zip(b.iter())
            .map(|(x, y)| x - y)
            .collect();
        
        if n > 0 {
            self.merge_sort(&mut c, 0, n - 1, d);
        }
        
        self.count
    }
}

fn main() {
    let mut sol = Solution::new();
    let nums1 = vec![3, -1];
    let nums2 = vec![-2, 2];
    let diff = -1;
    
    let result = sol.number_of_pairs(nums1, nums2, diff);
    println!("{}", result);
}